### ISSUE TITLE
Spectre Variant 1 - FENCE check

### ISSUE THREAD

**Issue by AlexeiMQ** [speed47/spectre-meltdown-checker#128]
V.0.34 Line 1280 - the script counts LFENCE (load serialization) only.
Seems like some kernels (for example Amazon Linux, Amazon Linux 2) use MFENCE (load/store serialization) instead.

If this assumption is correct:

1. replace Line 1280 to:
nb_lfence=$(objdump -d "$vmlinux" | grep -w -B1 'lfence|mfence' | grep -Ewc 'jmp|jne|je')

2. the following lines have to be added after Line 1315:
elif [ "$msg" = "Vulnerable" ] && [ "$v1_lfence" = 1 ]; then
			pvulnstatus $cve OK "SOME MESSAGE THERE"

**Comment by knweiss** [comment]
It would be interesting to know the sources and difference of all the various Spectre v1 mitigation patches...

For the record: Here are the official Intel and AMD whitepapers regarding Spectre mitigation which list the recommended instructions:

- AMD: [Managing Speculation on AMD Processors](https://developer.amd.com/wp-content/resources/Managing-Speculation-on-AMD-Processors.pdf) (pdf). This is the URL to the **updated version**. The original release had several mistakes.
- Intel: [Speculative Execution Side Channel Mitigations](https://software.intel.com/sites/default/files/managed/c5/63/336996-Speculative-Execution-Side-Channel-Mitigations.pdf) (pdf)

**Comment by speed47** [comment]
Do we have evidence that the Amazon kernel has been patched for variant 1 ?
I grabbed a copy of their 4.9.62-10.57 and 4.9.76-38.79 kernels. The first one is dated early December and is not patched for any variant as expected. The second one, which is the most recent to date, is patched for variant 2 (retpoline) and meltdown (PTI). The script correctly finds out. But I see no known mitigation for variant 1.
Do you have evidence or a statement on their part saying that they did patch their 4.9.76 for variant 1 ?

**Comment by knweiss** [comment]
Regarding `lfence` / `mfence`: Looking at the [latest Spectre variant 1 mitigation patchset](https://lkml.org/lkml/2018/1/29/960) for 4.16 I noticed that these instructions will be patched in at boot time using the "alternatives" mechanism:

- [PATCH v6 04/13 x86: introduce barrier_nospec](https://lkml.org/lkml/2018/1/29/969)

I.e. if I'm not missing something you can't necessarily find them in the kernel's `objdump` disassembly.

**Comment by speed47** [comment]
@knweiss For this patch, looking for lfence opcodes won't work because as you said, this'll be introduced with the `alternative` mechanism which an "empty old opcode" (i.e. nothing to grep for). The lfence opcodes will be present in the `.altinstr_replacement` section of the kernel binary, so technically there will be more lfence there than in a kernel without this patch. However lfence has also been used for other purposes, and in my tests I found some old kernels with a LOT of lfence there, so this wouldn't be a proper way to detect the patch.
But I found an other way, and this check has actually already been implemented in the master branch. I look for the presence of `array_index_mask_nospec()`.
The problem for now is distros backporting early versions of the patch that just have the lfence added a bit everywhere in the code without anything else. That's the heuristic check of variant1 I implemented but it has some false positives/negatives...
